### PR TITLE
fix(ec2): private DNS for custom endpoints has incorrect default

### PR DESCRIPTION
**Pull request by flemjame-at-amazon**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

Currently if I create an InterfaceVpcEndpoint to a non-AWS service, I must set the privateDnsEnabled field to be false, like so:
```java
InterfaceVpcEndpoint vpce = new InterfaceVpcEndpoint(construct, "VPCe", InterfaceVpcEndpointProps.builder()
            // Private DNS is only supported by AWS services
            .privateDnsEnabled(Boolean.FALSE)
            .vpc(vpc)
            .service(providerService)
            .build());
```

This is because the default is "true", which works for AWS services but not for custom hosted services. This change exposes a privateDnsDefault for different IInterfaceVpcEndpointService implementations, and VpcEndpoint will use that as the default if privateDnsEnabled is not specified. This way, I could create a service like so, without having to specify private DNS settings:

```java
    // Here I create a VPC endpoint to a custom hosted service
    IInterfaceVpcEndpointService providerService = new InterfaceVpcEndpointService(serviceName, servicePort);

    InterfaceVpcEndpoint vpce = new InterfaceVpcEndpoint(construct, "VPCe", InterfaceVpcEndpointProps.builder()
            .vpc(vpc)
            .service(providerService)
            .build());
```

This should break no existing functionality:

- Users overriding privateDnsEnabled
    - Will continue to have their setting applied
- Users not overriding privateDnsEnabled
    - If they are using InterfaceVpcEndpointAwsService, then the default associated with InterfaceVpcEndpointAwsService will be applied, which is true, which matches the current behavior
    - Users cannot use an InterfaceVpcEndpoint without overriding privateDnsEnabled, because only AWS services support private DNS

## Commit Message
fix(ec2): private DNS for custom endpoints has incorrect default

Currently if I create an InterfaceVpcEndpoint to a non-AWS service, I must set the privateDnsEnabled field to be "false". This is because the default is "true", which works for AWS services but not for custom hosted services. This change exposes a privateDnsDefault for different IInterfaceVpcEndpointService implementations, and VpcEndpoint will use that as the default if privateDnsEnabled is not specified. This way, I could create a service without having to specify private DNS settings.

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts [hunk 1]
```diff
@@ -427,16 +445,21 @@ export class InterfaceVpcEndpoint extends VpcEndpoint implements IInterfaceVpcEn
 
     const subnets = props.vpc.selectSubnets({ ...props.subnets, onePerAz: true });
     const subnetIds = subnets.subnetIds;
-
-    const endpoint = new CfnVPCEndpoint(this, 'Resource', {
-      privateDnsEnabled: props.privateDnsEnabled !== undefined ? props.privateDnsEnabled : true,
+    const endpointProps = {
+      privateDnsEnabled: true,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: I don't like the mutable props object you're using here. Why not simply:

```
new CfnVPCEndpoint(this, 'resource', {
  privateDnsEnabled: props.privateDnsEnabled ?? props.service.privateDnsDefault ?? true,
  // ...
});
```

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts [hunk 2]
```diff
@@ -201,6 +201,11 @@ export interface IInterfaceVpcEndpointService {
    * The port of the service.
    */
   readonly port: number;
+
+  /**
+   * Whether Private DNS is supported by default.
+   */
+  readonly privateDnsDefault?: boolean;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: Why is this optional?

File: package.json [hunk 3]
```diff
@@ -14,6 +14,7 @@
   },
   "devDependencies": {
     "conventional-changelog-cli": "^2.0.31",
+    "eslint-plugin-node": "^11.0.0",
```
Comment (rix0rrr) [inline package.json]: Why?

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts [hunk 4]
```diff
@@ -218,9 +223,15 @@ export class InterfaceVpcEndpointService implements IInterfaceVpcEndpointService
    */
   public readonly port: number;
 
+  /**
+   * Whether Private DNS is supported by default.
+   */
+  public readonly privateDnsDefault?: boolean;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: Value can be assigned here directly, doesn't need to wait for constructor.

Just write:

```
  public readonly privateDnsDefault?: boolean = false
```

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts [hunk 5]
```diff
@@ -279,9 +290,15 @@ export class InterfaceVpcEndpointAwsService implements IInterfaceVpcEndpointServ
...
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: Same comment here.

File: package.json (hunk 3 above)
Comment (flemjame-at-amazon) [reply package.json]: Unintentional change, will back it out

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts (hunk 1 above)
Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: I've never seen '??' before -- did you mean a different operator?

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts (hunk 2 above)
Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: I made this optional for backwards compatibility -- any existing CDK code specifying a service like:

```typescript
  const interfaceVpcEndpoint = new ec2.InterfaceVpcEndpoint(stack, 'InterfaceEndpoint', {
    vpc,
    service: {
      name: 'com.amazonaws.us-west-2.workspaces',
      port: 80
    }
  });
```

Will get compilation errors if I don't make the field optional:

```
Property 'privateDnsDefault' is missing in type '{ name: string; port: number; }' 
but required in type 'IInterfaceVpcEndpointService'.
```

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts (hunk 4 above)
Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: Done

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts (hunk 5 above)
Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: Done

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: I see. That's not how you're supposed to supply that value though, but I get it.

File: package-lock.json [hunk 6]
```diff
@@ -0,0 +1,111 @@
+{
```
Comment (jogold) [inline package-lock.json]: @rix0rrr there's a `package-lock.json` committed here...

Comment (rix0rrr) [reply package-lock.json]: Ahh doh.

**Pull request merged**